gateway/oauth_manager_test.go

```diff
@@ -374,6 +376,87 @@ func TestAPIClientAuthorizeAuthCode(t *testing.T) {
 	})
 }
 
+func TestAPIClientAuthorizeRequestWithPKCE(t *testing.T) {
+	ts := StartTest()
+	defer ts.Close()
+
+	spec := LoadAPI(buildTestOAuthSpec(true))[0]
+
+	pID := CreatePolicy(func(p *user.Policy) {
+		p.ID = "TEST-4321"
+		p.AccessRights = map[string]user.AccessDefinition{
+			"test": {
+				APIID: "test",
+			},
+			"abc": {
+				APIID: "abc",
+			},
+		}
+	})
+
+	var redirectURI string
+	// If separator is not set that means multiple redirect uris not supported
+	if config.Global().OauthRedirectUriSeparator == "" {
+		redirectURI = "http://client.oauth.com"
+
+		// If separator config is set that means multiple redirect uris are supported
+	} else {
+		redirectURI = strings.Join([]string{"http://client.oauth.com", "http://client2.oauth.com", "http://client3.oauth.com"}, config.Global().OauthRedirectUriSeparator)
+	}
+	testClient := OAuthClient{
+		ClientID:          authClientID,
+		ClientSecret:      "",
+		ClientRedirectURI: redirectURI,
+		PolicyID:          pID,
+		MetaData:          map[string]interface{}{"foo": "bar", "client": "meta"},
+	}
+	spec.OAuthManager.OsinServer.Storage.SetClient(testClient.ClientID, "org-id-1", &testClient, false)
+
+	t.Run("Client authorize token request with PKCE and no code challenge", func(t *testing.T) {
+		param := make(url.Values)
+		param.Set("response_type", "code")
+		param.Set("client_id", authClientID)
+		param.Set("key_rules", keyRules)
+
+		headers := map[string]string{
+			"Content-Type": "application/x-www-form-urlencoded",
+		}
+
+		_, _ = ts.Run(t, test.TestCase{
+			Path:      "/APIID/tyk/oauth/authorize-client/",
+			AdminAuth: true,
+			Data:      param.Encode(),
+			Headers:   headers,
+			Method:    http.MethodPost,
+			Code:      http.StatusForbidden,
+		})
+	})
+
+	t.Run("Client authorize token request with PKCE and code_challenge", func(t *testing.T) {
+
+		challenge := "12345678901234567890123456789012345678901234567890"
+
+		param := make(url.Values)
+		param.Set("response_type", "code")
+		param.Set("client_id", authClientID)
+		param.Set("key_rules", keyRules)
+		param.Set("code_challenge", challenge)
```

Comment:
Would like to see a test with different `code_challenge_method`s. e.g. Plain / S256
Would like to see examples of failed auth using incorrect code_verifier (plan & s256)

Reply:
Yep thats easy to add and is handled in the underlying lib

Reply:
What if there are 2 gateways in the cluster, and the authorize request happens on one GW, and the token request happens on another. Does the code_challenge get stored against the authorize request in Redis or something?

Reply:
The code challenge lives on the client so the same as current behaviour basically.